Author: Agustin Martin Domingo <agmartin@debian.org>
Description: Revert (sgml-lexical-context) use in flyspell (bzr#98607 and bzr#98656)
    
Use of (sgml-lexical-context) has some problems:
            
 * It is not available under XEmacs.
 
 * There is some obscure conflict with psgml (probably a psgml problem)
   that sometimes causes errors of the form:
                            
   Error in post-command-hook: (wrong-type-argument stringp nil)
   sgml-lexical-context: Wrong type argument: stringp, nil
                                              
   This happens even in emacs-snapshot when used original code with
   (sgml-lexical-context) together with psgml as preferred sgml mode.
   psgml provides 'sgml-mode, but has no `sgml-lexical-context'
   definition. Either if I manually load 'sgml-mode' or trust autoload
   at some points I sometimes get errors above.
                                                                                
   Function seems loaded since describe-function sgml-lexical-context
   works, and I can even explicitly evaluate it in the buffer (and
   seems to evaluate correctly).
   
* Is way slower than an own dedicated regexp.

--- a/support/emacsen/flyspell.el
+++ b/support/emacsen/flyspell.el
@@ -394,9 +394,15 @@ Tag and attribute names are not spell ch
 
 String values of attributes are checked because they can be text
 like <img alt=\"Some thing.\">."
-
-  (not (memq (car (sgml-lexical-context))
-	     '(tag pi))))
+  (not (save-excursion
+	 (or (and (looking-at "[^<\n]*>")          ;; Has unmatched > to the right
+		  (not (looking-at "[^=>\n]*\""))) ;; Not inside a string in a tag
+	     (and (ispell-looking-back "<[^>\n]*"  ;; Has unmatched < to the left
+				       (line-beginning-position))
+		  (not (looking-at "[^=\n]*\"")))  ;; Not inside a string
+	     (and (looking-at "[^&\n]*;")          ;; An SGML entity
+		  (ispell-looking-back "&[^;\n]*"
+				       (line-beginning-position)))))))
 
 ;;*---------------------------------------------------------------------*/
 ;;*    Programming mode                                                 */
